src/NextGen/Gateways/Stripe/NextGenStripeGateway/Webhooks/Listeners/InvoicePaymentSucceeded.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+namespace Give\NextGen\Gateways\Stripe\NextGenStripeGateway\Webhooks\Listeners;
+
+use Give\Donations\Models\Donation;
+use Give\Donations\Models\DonationNote;
+use Give\Donations\ValueObjects\DonationStatus;
+use Give\Donations\ValueObjects\DonationType;
+use Give\Framework\Exceptions\Primitives\Exception;
+use Give\Framework\Support\Facades\DateTime\Temporal;
+use Give\Framework\Support\ValueObjects\Money;
+use Give\NextGen\Gateways\Stripe\NextGenStripeGateway\NextGenStripeGateway;
+use Give\Subscriptions\Models\Subscription;
+use Give\Subscriptions\ValueObjects\SubscriptionStatus;
+use Stripe\Event;
+use Stripe\Invoice;
+
+/**
+ * @unreleased
+ */
+class InvoicePaymentSucceeded
+{
+    use StripeWebhookListenerRepository;
+
+    /**
+     * Processes invoice.payment_succeeded event.
+     *
+     * Occurs whenever an invoice payment attempt succeeds.
+     *
+     * @see https://stripe.com/docs/api/events/types#event_types-invoice.payment_succeeded
+     *
+     * @unreleased
+     *
+     * @return void
+     * @throws \Exception
+     */
+    public function __invoke(Event $event)
+    {
+        try {
+            $this->processEvent($event);
+        } catch (Exception $exception) {
+            $this->logWebhookError($event, $exception);
+        }
+
+        exit;
+    }
+
+    /**
+     * @unreleased
+     * @throws \Exception
+     */
+    public function processEvent(Event $event)
+    {
+        /* @var Invoice $invoice */
+        $invoice = $event->data->object;
+
+        $subscription = give()->subscriptions->queryByGatewaySubscriptionId($invoice->subscription)->get();
+
+        // only use this for next gen for now
+        if (!$subscription || !$this->shouldProcessSubscription($subscription)) {
+            return;
+        }
+
+        if ($initialDonation = give()->donations->getByGatewayTransactionId($invoice->payment_intent)) {
+            $this->handleInitialDonation($initialDonation);
+        } else {
+            if ($this->shouldCreateRenewal($subscription)) {
+                $subscription = $this->handleRenewal($subscription, $invoice);
+            }
+
+            if ($this->shouldEndSubscription($subscription)) {
+                $this->cancelSubscription($subscription);
+
+                $this->handleSubscriptionCompleted($subscription);
+            }
+        }
+    }
+
+    /**
+     * @unreleased
+     */
+    protected function shouldCreateRenewal(Subscription $subscription): bool
+    {
+        $billTimes = $subscription->installments;
+        $totalPayments = count($subscription->donations);
+
+        return $subscription->status->isActive() && (0 === $billTimes || $totalPayments < $billTimes);
```

Comment:
Is the number of installments not set in the subscription when it is created in Stripe? It just seems odd that we are manually checking that here. Just confirming that this is how it is supposed to be working (for my own understanding).